[Chee Jer Hong] iP

jhchee18 opened the pull request:
# RyanBot
> “Newton’s third law. You gotta leave something behind.” – Cooper, Interstellar ([source](https://kidadl.com/quotes/top-interstellar-quotes-about-the-beauty-and-power-of-the-universe))

RyanBot makes sure you don't miss your important _Tinder_ dates. :wink: It's,


- text-based
- not hard to learn
- ~~FAST~~ SUPER DUPERRRRR FAST to use :dash:
- All you need to do is,


1. download it the [v0.1 release](https://github.com/jhchee18/ip/releases/tag/A-Jar).
2. double-click it. :point_left::point_left:
3. add your tasks.
4. let it manage your tasks for you :grin:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Duke {
    public static void main(String[] args) {
        new Duke("duke.txt").run();
    }
```

Commit:
Add Gradle support

jhchee18 committed:
Level 1. Greet, Echo, Exit

jhchee18 committed:
Level 2: enable Add Task and List

jhchee18 committed:
Level 3: added Mark and Unmark methods to Task

jhchee18 committed:
Level 4: Added child classes (ToDo, Event, Deadline) to Parent Class (Task)

jhchee18 committed:
A-TextUiTesting: Input commands, checked expected and actual to be the same

jhchee18 committed:
Level 5: Handled 4 errors (ArrayIndexOutOfBound, IndexOutOfBound, InvalidCommand and EmptyTask)

jhchee18 committed:
Level 6: support for deleting tasks from the list

jhchee18 committed:
Improve last week code design

Remove Action and its respective sub-classes to change the implementation into just printing Strings

jhchee18 committed:
Level 7: Add Save feature to write List into txt file

Create new Storage class to handle creation of new File and writing into the file

jhchee18 committed:
Level-8: Added new LocalDate and LocalTime instances in Task

Convert String to LocalDate and LocalTime using new methods in Duke. Store deadline date and time in Task objects, instead of String.

jhchee18 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Level-8: Added new LocalDate and LocalTime instances in Task

# Conflicts:
#	src/main/java/Duke.java

jhchee18 committed:
A-MoreOOP: Create Command class and its child classes

Command classes are to handle different type of user input and convert them into different type of command sub-classes

jhchee18 committed:
A-MoreOOP: Modify the code design to adapt to OOP paradigm

More classes like Parser/Ui are created to make the code design more OOP

jhchee18 committed:
A-Packages: Divide classes into packages

Organize the classes into suitable java packages.e.g., duke.task, duke.command, duke

jhchee18 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

jhchee18 committed:
A-Gradle: Change mainClassName

To enable project to run in gradle

jhchee18 committed:
A-JUnit: Use JUnit Test to automate testing of different classes

Set up and implement JUnit to test out Parser, Storage, Todo, and Task tests

jhchee18 committed:
Add files for checkstyle

Configure build.gradle to use the two xml files added to perform checkstyle by following the style of SE-EDU

jhchee18 committed:
Checkstyle and fix styling

Follow the standard style recommended by SE-EDU except adding javadoc, which will be done later

jhchee18 committed:
Enable find task command

Users are now able to find task by searching the relavant keyword

jhchee18 committed:
Add JavaDoc comments

Add header comments to all non-private classes/methods.

jhchee18 committed:
Merge branch 'branch-A-JavaDoc'

jhchee18 committed:
Change styling and add JavaDoc

Checkstyle and change to adapt to styling for Parser and insert comment of JavaDoc for newly merged FindCommand

jhchee18 committed:
Setup JavaFX and make a simple GUI application with Hello World!

jhchee18 committed:
Design the layout of the Controls

jhchee18 committed:
Style the Controls to make the GUI fits the normal size.

jhchee18 committed:
Enable the GUI to echo the message of the user input in the TextField control

jhchee18 committed:
Add Dialog Boxes containing Label and Image, and add new package for frontend files

jhchee18 committed:
Add custom behavior (flipping Duke's DialogBox) to DialogBox

jhchee18 committed:
Rebuild the simple scho GUI Scene using FXML

jhchee18 committed:
Include Main class

jhchee18 committed:
Make the GUI able to adhere to the functionality of the backend code and improve the GUI

jchilling commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,192 @@
+    private Image user = new Image(this.getClass().getResourceAsStream("/images/remy.png"));
+    private Image duke = new Image(this.getClass().getResourceAsStream("/images/linguini.png"));
+
+    /**
+     * Constructs a Duke instance by default with no properties initialised
+     */
+    public Duke() {
+
+    }
```
Is this necessary?

jchilling reviewed (commented):
Looks good! 👍

jhchee18 committed:
Enable Bot to greet when initialize and farewell when terminate

Change the code design to adhere to the current GUI Fxml structure to
enable greet.

jhchee18 committed:
Disable TextField and Button after exiting

Stop allowing user to type in or enter new command.

jhchee18 committed:
Enhance GUI to display the pictures in circle frames

To make the GUI to look better and more related to real life product

jhchee18 committed:
Fix styling by following the checkstyleMain

YZTangent commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,32 @@
+import java.io.IOException;
+
+import duke.frontend.MainWindow;
+import javafx.application.Application;
```
I think java library imports should be placed above duke. imports according to checkstyle rules.

xiaobill8 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,65 @@
+     * Replace this stub with your completed method.
+     */
+    public String getResponse(String input) {
+        Command c = Parser.parse(input);
```
I like how you encapsulated the entire command into a Command object

xiaobill8 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,119 @@
+            String taskName = "";
+            String dateString = "";
+            String timeString = "";
+            for (int i = 1; i < parts.length; i++) {
```
Perhaps you could let Java's Datetime packages such as DateTimeFormatter and LocalDate.parse() handle DateTime parsing, so that you do not have to do so manually

xiaobill8 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public Deadline(String taskName, boolean markDone, LocalDate date, LocalTime time) {
+        super(taskName, markDone, date, time);
```
Good use of super(), since Deadline and Event both contains DateTime

xiaobill8 reviewed (commented):
Good use of OOP Principles overall, especially liked how you handled the Task constructors concisely.

jhchee18 committed:
Configure to enable assert feature

To be ready to document important assumptions that should hold at
various points in the code.

Testing of assert false in Parser class (now commented to ensure the
code runnable).

jhchee18 committed:
Add assertions in MainWindow.java

Ensure the program execute with correct assumptions.

jhchee18 committed:
DukeException: collate all exceptions thrown

Restructure all the exceptions in the Command sub-classes (e.g.
AddCommand, FindCommand, MarkCommand, etc.) to make the code more
understandable.

Converting all the exceptions to DukeException with a message to be
display later on.

Let's catch all the exceptions appear then throw a new DukeException
with a new message embed. The DukeException is then handled on another
class.

jhchee18 committed:
Refactor if-else statement into switch-case statement

Improve code quality in classes AddCommand and Parser to make them more
organized.

jhchee18 committed:
Use default branch in if-else statement

if and else-if in MarkCommand now is explicitly stated. Else case
is use to detect error.

YZTangent commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,66 @@
+        String[] parts = fullCommand.split(" ", 0);
+        String command = parts[0];
+        String mainCommand = "";
+        if (command.equals("todo") || command.equals("deadline") || command.equals("event")) {
+            mainCommand = "Add";
+        } else if (command.equals("delete")) {
+            mainCommand = "Delete";
+        } else if (command.equals("bye")) {
+            mainCommand = "Exit";
+        } else if (command.equals("mark") || command.equals("unmark")) {
+            mainCommand = "Mark";
+        } else if (command.equals("list")) {
+            mainCommand = "List";
+        } else if (command.equals("find")) {
+            mainCommand = "Find";
```
I think this set of if else statements can also be changed to switch statements too, with fallthrough behaviour for the first 3 or cases :)

YZTangent commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,66 @@
+            mainCommand = "Find";
+        }
+
+        switch (mainCommand) {
+        case "Delete":
+            return new DeleteCommand(Integer.parseInt(parts[1]) - 1);
+        case "Exit":
+            return new ExitCommand();
+        case "Add":
+            return new AddCommand(fullCommand);
+        case "Mark":
+            return new MarkCommand(command, Integer.parseInt(parts[1]) - 1);
+        case "List":
+            return new ListCommand();
+        case "Find":
+            return new FindCommand(fullCommand.substring(5));
+        default:
+            return new EmptyCommand();
+        }
```
Great job formatting the switch statements with correct indentation!

YZTangent reviewed (commented):
The PR is looking great! :)

jhchee18 committed:
Stop tracking duke.txt

jhchee18 committed:
Ignore duke.txt

jhchee18 committed:
Refactor the TaskList to achieve encapsulation

Reduce the occurrence of using the method getList() and delegate the
work to all the other methods in TaskList (e.g. setTask, getSize, etc.)

jhchee18 committed:
Reduce deep nesting in FindCommand

Extract out method to checkExistenceOfKeyword.

jhchee18 committed:
Merge pull request #1 from jhchee18/A-Assertions

A-Assertions: Add assertions in MainWindow.java

jhchee18 committed:
Merge branch 'master' into A-CodeQuality

# Conflicts:
#	data/duke.txt
#	src/main/java/duke/Parser.java
#	src/main/java/duke/command/EmptyCommand.java
#	src/main/java/duke/frontend/MainWindow.java

jhchee18 committed:
Add new reminder feature

Reminder class is created to generate reminder to remind user the
nearest coming task ahead.

It provides details such as countdown to the task from now.

jhchee18 committed:
Format reminder to become user readable

Convert Period to readable String so that user can understand and
add additional constraint to reminder (the nearest coming reminder must
not be done by the user).

jhchee18 committed:
Enhance GUI to have better appearance

Change background colors, resize image and add padding to dislogbox

jhchee18 committed:
Add User Guide

jhchee18 committed:
Rename jar file

jhchee18 committed:
Add initialization of duke.txt in root folder

jhchee18 committed:
Enhance display message if task list is empty

jhchee18 committed:
Fix bug to display full message without ellipsis

jhchee18 committed:
Add user guide

jhchee18 committed:
Add representative screenshot to showcase Ui

jhchee18 committed:
Add initialization of duke.txt

If run of new device, the new duke.txt created will be added with a
few dummy tasks to let user to test them out

jhchee18 committed:
Fix styling

jhchee18 committed:
Rename duke.txt to mytasklist.txt

jhchee18 committed:
Change file name in gitignore

jhchee18 committed:
Change name of tasklist file in some relevant files

jhchee18 committed:
Add DukeException to be thrown in testParser

jhchee18 committed:
Modify exception message to direct user to User Guide website

jhchee18 committed:
Rename boolean variable

jhchee18 committed:
Remove dead code

jhchee18 committed:
Reduce method length by using SLAP

jhchee18 committed:
Reduce method length by using SLAP

jhchee18 committed:
Improve code quality

jhchee18 committed:
Update User Guide with added sample GUI and new reminder feature

jhchee18 committed:
Update User guide

jhchee18 committed:
Update user guide

jhchee18 committed:
Add new HelpCommand class to provide guide to user

HelpCommand prompts User Guide website in browser.

jhchee18 committed:
Update user guide

jhchee18 committed:
Handle command with wrong format

jhchee18 committed:
Handle assertion error

jhchee18 committed:
Allow console to show error message

jhchee18 committed:
Modify task representation format

jhchee18 committed:
Fix bug in reminder to eliminate negative countdown

jhchee18 committed:
Include JUnit test to test out the countdown method in Reminder

jhchee18 committed:
Fix style

jhchee18 committed:
Added detailed error message

jhchee18 committed:
Updated tasklist
